engine: return target bytes consumed by scans

tbg opened the pull request:
To thread TargetBytes into the KV API, we will need to teach DistSender
to compute the remaining TargetBytes for each Range it's scanning. This
means it has to know how much was consumed by each individual response,
and so MVCC should return that information.

It also gives some additional test coverage, which in itself is already
enough of a reason to do it.

Release note: None

tbg committed:
engine: extract MVCCScanResult

MVCCScan and its brethren had a long list of return values already. I'm
going to be adding yet another one (the number of bytes represented in
the result), so the time had come for an aggregating struct.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44641)
<!-- Reviewable:end -->

tbg committed:
engine: return target bytes consumed by scans

To thread TargetBytes into the KV API, we will need to teach DistSender
to compute the remaining TargetBytes for each Range it's scanning. This
means it has to know how much was consumed by each individual response,
and so MVCC should return that information.

It also gives some additional test coverage, which in itself is already
enough of a reason to do it.

Release note: None

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44641#-:-M-BGdQqD_w89ThpB_Gq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @tbg)

___
*[c-deps/libroach/include/libroach.h, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44641#-M-BEg1w3ZVyb0lwH9HQ:-M-BEg1w3ZVyb0lwH9HR:b-5oaoj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4105a84126437c2afe2358df2e657a5a624511e1/c-deps/libroach/include/libroach.h#L354)):*
> ```objc
>   int32_t count;
>   // bytes is the number of bytes (as measured by TargetSize) in bufs.
>   int32_t bytes;
> ```

No results larger than 2GB? 


<!-- Sent from Reviewable.io -->

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44641#-:-M-BYvjlEYIHoQPvIEI-:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44641)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @petermattis)

___
*[c-deps/libroach/include/libroach.h, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44641#-M-BEg1w3ZVyb0lwH9HQ:-M-EKEkaAVmNXTVTBxOI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4105a84126437c2afe2358df2e657a5a624511e1/c-deps/libroach/include/libroach.h#L354)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

No results larger than 2GB? 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44641)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @itsbilal and @petermattis)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=petermattis,itsbilal
TFTRs!
